_drafts/linux-coredumps-ii.md

```diff
@@ -0,0 +1,246 @@
+to be an acceptable tradeoff, however, as most crashes can be debugged with just
+the stacktrace and stack allocated variables.
+
+The second tradeoff is that we're limiting the size of each stack. This does
+limit the depth of the stacktrace, but in general we have observed two things
+with coredumps. Typically the stacks are not that deep, and for those that are,
+only the top few frames are actually of any interest. This allows us to limit
+the size here, without affecting the debugging in the vast majority of cases.
+
+### Finding The Stacks
+
+If we want to both only capture the stacks, as well as limit the number of bytes
+we capture from each stack, we must find them. Recall from our previous article
+how a coredump represents mapped memory as a collection of `PT_LOAD` segments.
+If we want to find the `PT_LOAD` segment that contains the stack of the thread
+we're looking at, we need to know its PC. With the current PC of the thread, we
+can search through all segments and see which segment it falls into.
+
+But how do we get the current PC for each thread? Recall in our last article we
+briefly touched on all the different types of `PT_NOTE` segments that existed.
+One note in particular, `NT_PRSTATUS` is of interest here. This note contains
+information about each process. Here[^prstatus] is the layout of the note
+description:
+
+```c
+struct elf_prstatus_common
+{
+ struct elf_siginfo pr_info; /* Info associated with signal */
+ short pr_cursig;  /* Current signal */
+ unsigned long pr_sigpend; /* Set of pending signals */
+ unsigned long pr_sighold; /* Set of held signals */
+ pid_t pr_pid;
+ pid_t pr_ppid;
+ pid_t pr_pgrp;
+ pid_t pr_sid;
+ struct __kernel_old_timeval pr_utime; /* User time */
+ struct __kernel_old_timeval pr_stime; /* System time */
+ struct __kernel_old_timeval pr_cutime; /* Cumulative user time */
+ struct __kernel_old_timeval pr_cstime; /* Cumulative system time */
+};
+
+struct elf_prstatus
+{
+ struct elf_prstatus_common common;
+ elf_gregset_t pr_reg; /* GP registers */
+ int pr_fpvalid;  /* True if math co-processor being used.  */
+};
+```
+
+Of interest here is the `pr_reg` struct memmber. This has all of the general
+purpose registers at the time of the crash. From here we can grab the PC of the
+thread, and search through all of the mapped memory regions. Once we find the
+mapped memory region in question we can grab the last N bytes before the PC.
+
+Now we have just the memory from each thread! If we tried to load this into GDB,
+however it wouldn't work. That's because we're missing some vital bits of debug
+information. Let's take a look what these sections are, and how we can identify
+them.
+
+### `r_debug` - Debug Rendevous Structure
+
+When we write a Linux program, we're typically pulling in more code than what we
+directly compile in. These bits of code are loaded in at runtime, and are
+referred to as dynamically linked. Examples of these are `libc`, `libopenssl`,
```

Comment:
```suggestion
Dynamic linking allows programs to call code from shared libraries on the system. For example. this can allow a single `libopenssl` binary to handle cryptography for a variety of applications without the full library binary needing to be included in each one. 
```